data_model/src/domain.rs

```diff
@@ -108,6 +126,42 @@ impl Domain {
             logo: None,
         }
     }
+
+    /// Returns a reference to the account corresponding to the account id.
+    pub fn get_account(&self, account_id: &<Account as Identifiable>::Id) -> Option<&Account> {
+        self.accounts.get(account_id)
+    }
+
+    /// Returns a mutable reference to the account corresponding to the account id.
+    pub fn get_account_mut(
```

Comment:
most if not all of the `xxx_mut` functions are used or should be used only in `iroha_core`. If I encapsulate the access to fields of the structures, I'm forced to expose these methods as public so that `iroha_core` can access them and modify it's internal state. This, however, makes them available to the end user as well. I don't like this because user should only be concerned with creating data model structures and inspecting them when returned in queries (when returned in queries they are copied so there is no feat of messing up iroha internal state, nevertheless...). In this example, of domain with accounts, user should only be exposed fns such as `get_account`, `add_account`, `remove_account`, `accounts` but not the methods such as `get_account_mut` or `accounts_mut`

This can be tackled if we were to have both `core::data_model` and `user::data_model` and functions which convert between the two. One could say that it's reasonable to have representation layer separated from the implementation layer so that forms presented to the end user don't have to match the internal state. However, this seems like an overkill to me

FFI is a different matter and it's API doesn't have to include mutation methods. Though if we were to have both `core::data_model` and `user::data_model` FFI access would go like this: FFI function -> user::data_model -> core::data_model`. There is one more conversion in the process

Reply:
@appetrosyan I think leaving mutation methods accessible to the user is a more prudent choice here.